### ISSUE TITLE
URLSessionInstrumentation Hang Rate

### ISSUE THREAD

**Issue by Sherlouk** [open-telemetry/opentelemetry-swift#380]
Hey folks,

We've recently released OpenTelemetry with URL instrumentation into production, and noticed a sharp increase in our hang rate metric. The vast majority of the hang rate reports coming from Apple are being linked to the following functions:

https://github.com/open-telemetry/opentelemetry-swift/blob/main/Sources/Instrumentation/URLSession/URLSessionInstrumentation.swift#L50

https://github.com/open-telemetry/opentelemetry-swift/blob/main/Sources/Instrumentation/URLSession/InstrumentationUtils.swift#L9

Is there an expectation that setup is called on a non-main thread or is there perhaps something else going wrong here?

**Comment by nachoBonafonte** [comment]
Hi Sherlouk,

Yes URLSession instrumentation setup is quite time consuming, it currently iterates all classes to find urlsession delegates to add instrumentation. 

In latest versions it does in a Dispatch.concurrent method, but if the app links many classes it can still take some time. Which version of the library are you delivering?

In your case, that I think you control the whole stack, we could improve setup maybe adding the list of class names to instrument, it would definitely make it extremely faster. What do you think?

**Comment by Sherlouk** [comment]
We're currently on 1.2.1 so conscious an update is in order.

Also aware that Firebase and others, where they do similar instrumentation, have had no problems executing on the background thread - can you foresee any problems this might cause? If not, should it be the default when calling the initialiser?

Which classes, exactly, need to be picked up for instrumentation to work? We don't use any custom URLSessions, etc.

**Comment by nachoBonafonte** [comment]
> We're currently on 1.2.1 so conscious an update is in order.

That version already loads with the dispatch.concurrent method

> Also aware that Firebase and others, where they do similar instrumentation, have had no problems executing on the background thread - can you foresee any problems this might cause? If not, should it be the default when calling the initialiser?

Yes, initializing in a background thread is perfectly good approach, and will not take time from the main thread, I do it in my product without issues. 

Be aware, though, that if other network calls are made while the instrumentation is being initialized thread-race conditions can happen (Thread sanitizer can help pointing them if you have it enabled in some test that loads the app

> Which classes, exactly, need to be picked up for instrumentation to work? We don't use any custom URLSessions, etc.

Basically any class that implements URLSesssion delegate or similar. We could add an option for the configuration to avoid looking for them, which would also improve loading time a lot.

**Comment by Sherlouk** [comment]
> That version already loads with the dispatch.concurrent method

Noted, other improvements though that are worth grabbing.

> Yes, initializing in a background thread is perfectly good approach, and will not take time from the main thread, I do it in my product without issues.

Noted, is it worth updating documentation with notes around what was discussed here - it seems other apps may run into similar problems and can weigh up the pros/cons themselves.

> Basically any class that implements URLSesssion delegate or similar

Yeah we have nothing like that, so it seems it could be good to provide this as a configuration option - a solution which allows the user to provide their own or rely on the automatic detection. Of course providing your own could simply be an empty array thus eliminating a lot of the waiting around.
